Dart/src/com/jetbrains/lang/dart/analyzer/DartAnalysisServerService.java

```diff
@@ -331,6 +345,148 @@ public LibraryDependenciesResult(@Nullable final String[] libraries,
     }
   }
 
+  private void configureImportedLibraries(@NotNull final Collection<String> rootsToAddToLib) {
+    runInEventDispatchThread(new Runnable() {
+      @Override
+      public void run() {
+        ApplicationManager.getApplication().runWriteAction(new Runnable() {
+          public void run() {
+            doConfigureImportedLibraries(rootsToAddToLib);
+          }
+        });
+      }
+    });
+  }
+
+  private static void runInEventDispatchThread(final Runnable runnable) {
```

Comment:
Notification always comes in ServerResponseReaderThread, so it is safe to proceed without thread check.

More important is that there's no need in blocking ServerResponseReaderThread and waiting until EDT does project configuration. We do not need results immediately, so it is safe to call invokeLater() instead of invokeAndWait().

One more important knowledge is that (except some special cases) project model change should happen in non-modal state (instead of defaultState() that can be modal if, for example, Project Structure dialog is open).

And the last but not the least: you need to use index. That means that the code must not be run in Dumb mode when indexes are not ready.

Al these problems are solved by using `DumbService.getInstance(project).smartInvokeLater()`. I already use it in DartFileListener.